File: orgflow-ui/lib/auth/contractor-access.ts
/**
 * QC Spec 4.2.1 - Contractor persona access profile (limited permissions).
 * See docs/qc-spec/qc-personas-permissions.md §2.2.
 */

import type { QCPermission } from "@/lib/quality-issues/permissions";
import {
  hasQCPermission,
  resolveQCPermissions,
  resolveQCPersona,
  visibleIssueStatusesForRole,
} from "@/lib/quality-issues/permissions";

export const CONTRACTOR_GRANTED_PERMISSIONS = [
  "quality_issues:read",
  "quality_issues:remediate",
  "projects:read",
] as const satisfies readonly QCPermission[];

export const CONTRACTOR_DENIED_PERMISSIONS = [
  "quality_issues:write",
  "quality_issues:verify",
  "quality_portfolio:read",
  "field_reports:read",
  "field_reports:write",
  "field_reports:admin",
  "reports:read",
  "reports:write",
  "users:read",
  "users:write",
  "audit:read",
] as const satisfies readonly QCPermission[];

export function isContractorRole(role?: string | null): boolean {
  return resolveQCPersona(role) === "CONTRACTOR";
}

export function getContractorAccessProfile(role?: string | null) {
  const permissions = resolveQCPermissions(role);
  const visibleStatuses = visibleIssueStatusesForRole(role);

  return {
    persona: resolveQCPersona(role),
    permissions: [...permissions],
    visibleIssueStatuses:
      visibleStatuses === null ? null : [...visibleStatuses],
    canReadIssues: hasQCPermission(role, "quality_issues:read"),
    canRemediateIssues: hasQCPermission(role, "quality_issues:remediate"),
    canAccessFieldReports: hasQCPermission(role, "field_reports:read"),
    canAccessPortfolio: hasQCPermission(role, "quality_portfolio:read"),
  };
}

export function contractorHasLimitedAccess(role?: string | null): boolean {
  if (!isContractorRole(role)) {
    return false;
  }

  const profile = getContractorAccessProfile(role);

  return (
    profile.canReadIssues
    && profile.canRemediateIssues
    && !profile.canAccessFieldReports
    && !profile.canAccessPortfolio
    && profile.visibleIssueStatuses !== null
    && profile.visibleIssueStatuses.includes("OPEN")
    && profile.visibleIssueStatuses.includes("IN_REMEDIATION")
    && !profile.visibleIssueStatuses.includes("CLOSED")
  );
}

File: app/schemas/contractor_access.py
"""
QC Spec 4.2.1 - Contractor persona access profile (limited permissions).

See docs/qc-spec/qc-personas-permissions.md §2.2.
"""

from __future__ import annotations

from app.schemas.qc_permissions import (
    QCPersona,
    has_qc_permission,
    resolve_qc_permissions,
    resolve_qc_persona,
    visible_issue_statuses_for_role,
)

CONTRACTOR_GRANTED_PERMISSIONS: tuple[str, ...] = (
    "quality_issues:read",
    "quality_issues:remediate",
    "projects:read",
)

CONTRACTOR_DENIED_PERMISSIONS: tuple[str, ...] = (
    "quality_issues:write",
    "quality_issues:verify",
    "quality_portfolio:read",
    "field_reports:read",
    "field_reports:write",
    "field_reports:admin",
    "reports:read",
    "reports:write",
    "users:read",
    "users:write",
    "audit:read",
)


def contractor_can_access_field_reports(role: str | None) -> bool:
    return has_qc_permission(role, "field_reports:read")


def contractor_can_access_catalog(role: str | None) -> bool:
    return contractor_can_access_field_reports(role)


def is_contractor_role(role: str | None) -> bool:
    return resolve_qc_persona(role) == QCPersona.CONTRACTOR


def contractor_has_limited_access(role: str | None) -> bool:
    if not is_contractor_role(role):
        return False

    permissions = resolve_qc_permissions(role)
    visible = visible_issue_statuses_for_role(role)

    return (
        has_qc_permission(role, "quality_issues:read")
        and has_qc_permission(role, "quality_issues:remediate")
        and not has_qc_permission(role, "field_reports:read")
        and not has_qc_permission(role, "quality_portfolio:read")
        and visible is not None
        and all(
            granted in permissions for granted in CONTRACTOR_GRANTED_PERMISSIONS
        )
        and all(
            denied not in permissions for denied in CONTRACTOR_DENIED_PERMISSIONS
        )
    )
